### PR TITLE
Stadium List and Stadium View Beginnings

### LINKED ISSUE #16: Sort and display Stadiums
### Description:
As a baseball fanatic, I want to be able to see all of the stadiums along with the name of the teams that play there sorted by the typology of the stadium, chronology of the opening date of the stadium, and seating capacity of the stadiums so that I can plan my trip just the way I want to.

### Assumptions:
* The main window has been initialized
* The data for the stadiums has been loaded into the program

### Estimate:
* 13

### Tasks:
* Create a view for displaying the stadiums and their information
* Give the user the ability to navigate to a list of the stadiums sorted in the desired way
* Populate the view with the desired sorted list of stadiums
* If the user ever decides to sort in a different way, it should clear the old list and display a new one

### Tests:
* Verify that the all desired stadiums are loaded in to the view
* Verify that stadiums that are not desired are not loaded in along with the correct stadiums
* Verify that the stadiums are sorted in the desired fashion

### Definition of Done:
The user can access the list of stadiums along with the names of the teams that play at each stadium sorted in the desired fashion.

### PR DESCRIPTION
**Pull request by WakkaTrout**
Closes #16 

### Key features
* Added Stadium List Class
* Added Stadium List Item Class
* Added Stadium View Class (not finished)
* Completed sorts for the stadium list

### Future Improvements
* Finish adding extra functionality to Stadium View for specific sorts

### Notes
* When a row is clicked, it emits a stadium id for use in other parts of the program
* I put the vector for the views in main window so that I could see the stadium view (should be a copy of what Chris has in the Admin view branch)

### CODE REVIEW

File: GrayonSlam.pro [hunk 1]
```diff
@@ -53,13 +56,18 @@ HEADERS += \
     src/widgets/souvenirlistitem.hpp \
     src/utils/map.hpp \
     src/datastore/database.hpp \
-    src/utils/enumtools.hpp
+    src/utils/enumtools.hpp \
+    src/views/stadiumview.hpp \
+    src/widgets/stadiumlist.hpp \
+    src/views/stadiumview.hpp \
```
Comment (ChrisRzech) [inline GrayonSlam.pro]: Duplicate

File: src/widgets/stadiumlist.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+StadiumList::StadiumList(QWidget *parent) :
+    QWidget(parent)
+{
+    //creates new tree widget
+    m_listDisplay = new QTreeWidget(this);
+    m_listDisplay->resize(parent->size());
+    //populates the tree widget from the database
+    populateWidget(Database::getTeamsAndStadiums());
+    //sets up the stadium clicked signal
+    QObject::connect(m_listDisplay, SIGNAL(itemClicked(QTreeWidgetItem*, int)),this, SLOT(connectWidgetItemToStadium(QTreeWidgetItem*)));
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp]: Use Qt-5 style connects
```cpp
connect(m_listDisplay, &QTreeWidget::itemClicked,
        this, &StadiumList::connectWidgetItemToStadium);
```
(not sure if u can drop a parameter, test it out)

File: src/widgets/stadiumlist.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
+#include "stadiumlist.hpp"
+
+/**
+ * This constructor constructs a new StadiumList inside of the parent widget and populates it with
+ * the information from the database
+ * @param parent pointer to parent widget
+ */
+StadiumList::StadiumList(QWidget *parent) :
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp]: Didn't your momma teach you about delegating ctors?
```cpp
StadiumList::StadiumList(QWidget* parent)
    : StadiumList(Database::getTeamsAndStadiums(), parent)
{}
```

File: src/widgets/stadiumlist.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    populateWidget(stadiumList);
+    //sets up the stadium clicked signal
+    QObject::connect(m_listDisplay, SIGNAL(itemClicked(QTreeWidgetItem*, int)),this, SLOT(connectWidgetItemToStadium(QTreeWidgetItem*)));
+}
+
+/**
+ * setStadiumTeamList resets the tree widget and populates it with the information from the vector passed in
+ * @param stadiumList information to repopulate the tree widget with
+ */
+void StadiumList::setStadiumTeamList(const std::vector<std::pair<Team,Stadium>>& stadiumList)
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp]: Not sure you why have this function if it just calls `populateWidget()`, redundant

File: src/widgets/stadiumlist.hpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    /* Resets */
+    void setStadiumTeamList(const std::vector<std::pair<Team,Stadium>>&);
+
+    /* Destructor*/
+    virtual ~StadiumList();
+
+signals:
+    /* Clicked Signal*/
+    void StadiumClicked(int stadiumId);
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.hpp]: camelCasing
😢

File: src/widgets/stadiumlist.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+{
+    //clears old display
+    m_listDisplay->clear();
+    m_listDisplay->setColumnCount(10);
+    QStringList headers = { "Team Name", "League", "Stadium Name", "Location", "Date Opened",
+    "Seating Capacity", "Typology", "Roof Type", "Playing Surface", "Distance to Center Field"};
+    m_listDisplay->setHeaderLabels(headers);
+    for(std::pair<Team,Stadium> stadiumAndTeam : stadiumsAndTeams)
+    {
+        if(stadiumAndTeam.first.getId() != -1 && stadiumAndTeam.second.getId() != -1)
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp]: Let's allow invalid teams to be displayed, invalid stadiums shouldn't be displayed though

File: src/widgets/stadiumlist.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            tmpList.push_back(QString::fromStdString(stadiumAndTeam.second.getLocation()));
+            tmpList.push_back(QString::fromStdString(std::to_string(stadiumAndTeam.second.getYearOpened())));
+            tmpList.push_back(QString::fromStdString(commaSeparate(std::to_string(stadiumAndTeam.second.getSeatCap()))));
+            tmpList.push_back(QString::fromStdString(stadiumAndTeam.second.TYPOLOGY_STRING[stadiumAndTeam.second.typology]));
+            tmpList.push_back(QString::fromStdString(stadiumAndTeam.second.ROOF_STRING[stadiumAndTeam.second.roof]));
+            tmpList.push_back(QString::fromStdString(stadiumAndTeam.second.SURFACE_STRING[stadiumAndTeam.second.surface]));
+            tmpList.push_back(QString::fromStdString(commaSeparate(std::to_string(stadiumAndTeam.second.getCenterFieldDist()))));
+
+            //Constructs a new stadium list item and adds it to the tree widget
+            new StadiumListItem(m_listDisplay,stadiumAndTeam.second,tmpList);
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp]: WHERE'S MY VARIABLE


jk

File: src/widgets/stadiumlist.cpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+       emit StadiumClicked(itemS->getStadiumId());
+    }
+}
+
+/**
+ * populateWidget() takes a vector of stadiums and teams, converts their data to QStrings, and inserts
+ * them into the Tree Widget
+ * @param stadiumsAndTeams vector containing stadiums and teams
+ */
+void StadiumList::populateWidget(const std::vector<std::pair<Team,Stadium>>& stadiumsAndTeams)
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp:56]: Turn off sorting before adding new items

File: src/widgets/stadiumlist.hpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <QTreeWidget>
+#include "src/datastore/stadium.hpp"
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.hpp]: Too many includes, you only need to include `"src/datastore/database.hpp"`

File: src/widgets/stadiumlist.hpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+#include "src/datastore/team.hpp"
+#include "src/datastore/database.hpp"
+#include "src/utils/enumtools.hpp"
+#include "stadiumlistitem.hpp"
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.hpp:4]: include not needed in this file

File: src/widgets/stadiumlist.hpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * This will create a stadium list within its parent widget.
+ * Then it will add all the stadiums and teams from the vector to the list for displaying.
+ */
+class StadiumList : public QWidget
+{
+    Q_OBJECT
+public:
+    /* Constructors*/
+    explicit StadiumList(QWidget *parent = nullptr);
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.hpp]: Don't allow defaulting to `nullptr`

File: src/widgets/stadiumlist.hpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    explicit StadiumList(const std::vector<std::pair<Team,Stadium>>&, QWidget *parent = nullptr);
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.hpp]: Don't allow defaulting to `nullptr`

File: src/widgets/stadiumlistitem.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+    //compare distance to center field
+    else if(parent->sortColumn() == 9)
+    {
+        lessThan = m_centerFieldDist < otherItem.m_centerFieldDist;
+    }
+    //compare string and not a number
+    else
+    {
+        lessThan = QTreeWidgetItem::operator<(otherItem);
```
Comment (ChrisRzech) [inline src/widgets/stadiumlistitem.cpp:45]: :POGGERS:
explicit operator call, once in a life time of actual needed use

File: src/widgets/stadiumlistitem.hpp [hunk 14]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <QTreeWidgetItem>
+#include "src/datastore/stadium.hpp"
+#include "src/datastore/team.hpp"
```
Comment (ChrisRzech) [inline src/widgets/stadiumlistitem.hpp]: Include not needed

File: src/widgets/stadiumlistitem.hpp [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+class StadiumListItem : public QTreeWidgetItem
+{
+public:
+    /* Constructor */
+    StadiumListItem(QTreeWidget* parent, const Stadium&, const QStringList&);
+
+    /* Operator Overloads*/
+    virtual bool operator<(const StadiumListItem&);
```
Comment (ChrisRzech) [inline src/widgets/stadiumlistitem.hpp:13]: do not mark `virtual`, mark `override` instead

File: src/widgets/stadiumlist.hpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+    /* Resets */
+    void setStadiumTeamList(const std::vector<std::pair<Team,Stadium>>&);
+
+    /* Destructor*/
+    virtual ~StadiumList();
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.hpp]: do not mark `virtual`, use `override`

**Review by ChrisRzech — CHANGES_REQUESTED** [review]
nice

File: src/widgets/stadiumlist.cpp (hunk 4 above)
Comment (WakkaTrout) [reply src/widgets/stadiumlist.cpp]: When the vector was there, it has meaning, but I forgot to take it out

File: src/widgets/stadiumlist.cpp (hunk 8 above)
Comment (WakkaTrout) [reply src/widgets/stadiumlist.cpp:56]: done

File: src/widgets/stadiumlistitem.hpp (hunk 15 above)
Comment (WakkaTrout) [reply src/widgets/stadiumlistitem.hpp:13]: needs to be virtual for calls, but ancestor not virtual (so cannot use overload specifier https://en.cppreference.com/w/cpp/language/override)

Comment (ChrisRzech) [reply src/widgets/stadiumlistitem.hpp:13]: Oh didn't notice this isn't overriding the base class one, ok then

File: src/widgets/stadiumlist.cpp [hunk 17]
```diff
@@ -0,0 +1,127 @@
+#include "stadiumlist.hpp"
+
+/**
+ * This constructor constructs a new StadiumList inside of the parent widget and populates it with
+ * the information from the database
+ * @param parent pointer to parent widget
+ */
+StadiumList::StadiumList(QWidget* parent) :
```
Comment (ChrisRzech) [inline src/widgets/stadiumlist.cpp]: YO where's my delegating ctor?

**Review by yosmo78 — APPROVED** [review]
It's beautiful and I see nothing wrong with it

**Review by ChrisRzech — APPROVED** [review]
freaking nice 👍

**Pull request merged**